### PR TITLE
SAA-118 initial commit for new endpoint for retrieval of attendance records for a scheduled activity.

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
Adds new endpoint for retrieval of attendances for marking attendance:

`/scheduled-instances/{instanceId}/attendances`

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Attendance.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  @Schema(description = "The reason for attending or not")
+  val attendanceReason: AttendanceReason? = null,
+
+  @Schema(
+    description = "Free text to allow comments to be put against the attendance",
+    example = "Prisoner was too unwell to attend the activity."
+  )
+  var comment: String? = null,
+
+  val posted: Boolean,
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Attendance.kt:25]: @timharrison-moj is this is to say whether this has submitted/processed?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Attendance.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  val status: String? = null,
+
+  @Schema(description = "The amount in pence to pay the prisoner for the activity", example = "100")
+  val payAmount: Int? = null,
+
+  @Schema(description = "The bonus amount in pence to pay the prisoner for the activity", example = "50")
+  val bonusAmount: Int? = null,
+
+  val pieces: Int? = null
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Attendance.kt:42]: @timharrison-moj and this one?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/config/OpenApiConfiguration.kt [hunk 3]
```diff
@@ -24,4 +25,9 @@ class OpenApiConfiguration(buildProperties: BuildProperties) {
             .email("feedback@digital.justice.gov.uk")
         )
     )
+
+  @Bean
+  fun enableLocalTimePrimitiveType() {
+    PrimitiveType.enablePartialTime()
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/config/OpenApiConfiguration.kt:31]: Just trialing this to see if this answer Marks questions/issues with the way types are be presented via the Swagger docs.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstance.kt [hunk 4]
```diff
@@ -23,6 +28,10 @@ data class ScheduledInstance(
   @JoinColumn(name = "activity_schedule_id", nullable = false)
   val activitySchedule: ActivitySchedule,
 
+  @OneToMany(mappedBy = "scheduledInstance", fetch = FetchType.EAGER, cascade = [CascadeType.ALL], orphanRemoval = true)
+  @Fetch(FetchMode.SUBSELECT)
+  val attendances: MutableList<Attendance> = mutableListOf(),
+
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstance.kt:34]: Will this eagerly fetch ALL attendances, or just those for the time period requested? Just thinking that the list could get very long over time.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Attendance.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  val posted: Boolean,
+
+  @Schema(description = "The date and time the attendance was updated", example = "10/09/2023")
+  @JsonFormat(pattern = "dd/MM/yyyy HH:mm")
+  val recordedTime: LocalDateTime? = null,
+
+  @Schema(description = "The person whom updated the attendance", example = "10/09/2023")
+  val recordedBy: String? = null,
+
+  val status: String? = null,
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/model/Attendance.kt:34]: Status similar to NOMIS - SCH (scheduled or planned), CANC (cancelled), COMP (completed) - attendance received, EXP (expired) - no updates received within 3 days. We might need to add to these later, if we want to track whether they've been paid or not. For now SCH and COMP should be the ones to focus on please

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/ScheduledInstanceController.kt [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.web.bind.annotation.RestController
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.config.ErrorResponse
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.model.Attendance
+import uk.gov.justice.digital.hmpps.hmppsactivitiesmanagementapi.service.AttendancesService
+
+// TODO add pre-auth annotations to enforce roles when we have them
+
+@RestController
+@RequestMapping("/scheduled-instances", produces = [MediaType.APPLICATION_JSON_VALUE])
+class ScheduledInstanceController(private val attendancesService: AttendancesService) {
```
Comment (timharrison-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/ScheduledInstanceController.kt:23]: Just wondering if the controller should be at the scheduled instance level, as this is an internal construct? Controllers usually better at the business level e.g. All endpoints relating to attendance under `/attendance` and endpoints relating to activities under `/activity`. Ok for now, but thinking about logical grouping of endpoints for beta.

**Review by timharrison-moj — APPROVED** [review]
Couple of comments, but LGTM for now.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstance.kt (hunk 4 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/entity/ScheduledInstance.kt:34]: This is only attendances for that particular instance i.e. session.  So if there ware 10 allocations there will be 10 attendance records.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/ScheduledInstanceController.kt (hunk 6 above)
Comment (matt-crowson-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/ScheduledInstanceController.kt:23]: I think this warrants a bit of a conversation to be honest in how we would like to use the API.

**Pull request merged**